File: android/src/main/java/com/Smartlook/Smartlook/smartlook/SmartlookPlugin.java
package com.Smartlook.Smartlook.smartlook;

import android.webkit.WebView;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;
import io.flutter.plugin.common.MethodChannel.MethodCallHandler;
import io.flutter.plugin.common.MethodChannel.Result;
import io.flutter.plugin.common.PluginRegistry.Registrar;
import org.json.JSONObject;
import com.smartlook.sdk.smartlook.Smartlook;
import java.util.HashMap;
import com.google.gson.Gson;
import com.google.gson.reflect.TypeToken;
import java.lang.reflect.Type;

/** SmartlookPlugin */
public class SmartlookPlugin implements MethodCallHandler {
  /** Plugin registration. */
  public static void registerWith(Registrar registrar) {
    final MethodChannel channel = new MethodChannel(registrar.messenger(), "smartlook");
    channel.setMethodCallHandler(new SmartlookPlugin());
  }

  private Gson gson = null;

  @Override
  public void onMethodCall(MethodCall call, Result result) {
    if (call.method.equals("setupAndStartRecording")) {
      String apiKey = call.argument("key");
      int fps = call.argument("fps");
      Smartlook.setupAndStartRecording(apiKey, true, fps);
    } else if (call.method.equals("setup")) {
      String apiKey = call.argument("key");
      int fps = call.argument("fps");
      Smartlook.setup(apiKey, true, fps);
    } else if (call.method.equals("isRecording")) {
      result.success(Smartlook.isRecording());
    } else if (call.method.equals("startFullscreenSensitiveMode")) {
      Smartlook.startFullscreenSensitiveMode();
    } else if (call.method.equals("stopFullscreenSensitiveMode")) {
      Smartlook.stopFullscreenSensitiveMode();
    } else if (call.method.equals("isFullscreenSensitiveModeActive")) {
      result.success(Smartlook.isFullscreenSensitiveModeActive());
    } else if (call.method.equals("setReferrer")) {
      String referrer = call.argument("referrer");
      String source = call.argument("source");
      Smartlook.setReferrer(referrer, source);
    } else if (call.method.equals("trackNavigationEvent")) {
      String navKey = call.argument("key");
      int direction = call.argument("type");
      String viewDirection = "start";
      if (direction == 0) viewDirection = "start"; else viewDirection = "stop";
      Smartlook.trackNavigationEvent(navKey, "activity", viewDirection);
    } else if (call.method.equals("getDashboardSessionUrl")) {
      result.success(Smartlook.getDashboardSessionUrl());
    } else if (call.method.equals("startRecording")) {
      Smartlook.startRecording();
    } else if (call.method.equals("stopRecording")) {
      Smartlook.stopRecording();
    } else if (call.method.equals("flush")) {
      //Smartlook.flush();
    } else if (call.method.equals("removeAllGlobalEventProperties")) {
      Smartlook.removeAllGlobalEventProperties();
    } else if (call.method.equals("removeGlobalEventProperty")) {
      String propertyKey = call.argument("key");
      Smartlook.removeGlobalEventProperty(propertyKey);
    } else if (call.method.equals("setGlobalEventProperties")) {
      HashMap globalImmutableProperties = call.argument("map");
      boolean immutable = call.argument("immutable");

      if (globalImmutableProperties != null) {
        if (gson == null) {
          gson = new Gson();
        }
        Type gsonType = new TypeToken<HashMap>(){}.getType();
        String gsonString = gson.toJson(globalImmutableProperties,gsonType);
        Smartlook.setGlobalEventProperties(gsonString, immutable);
      }
    } else if (call.method.equals("startTimedCustomEvent")) {
      String eventKey = call.argument("key");
      Smartlook.startTimedCustomEvent(eventKey);
    } else if (call.method.equals("enableCrashlytics")) {
      boolean enabled = call.argument("enabled");
      Smartlook.enableCrashlytics(enabled);
    } else if (call.method.equals("enableWebviewRecording")) {
      boolean enabled = call.argument("enabled");
      if (enabled) {
        Smartlook.unregisterBlacklistedClass(WebView.class);
      } else {
        Smartlook.registerBlacklistedClass(WebView.class);
      }
    } else if (call.method.equals("trackCustomEvent")) {
      String eventKey = call.argument("key");
      HashMap eventMap = call.argument("map");

      if (eventKey != null && eventMap != null) {
        if (gson == null) {
          gson = new Gson();
        }
        Type gsonType = new TypeToken<HashMap>(){}.getType();
        String gsonString = gson.toJson(eventMap,gsonType);
        Smartlook.trackCustomEvent(eventKey, gsonString);
      } else if (eventKey != null) {
        Smartlook.trackCustomEvent(eventKey);
      }
    } else if (call.method.equals("setUserIdentifier")) {
      String identifyKey = call.argument("key");
      HashMap identifyMap = call.argument("map");

      if (identifyKey != null && identifyMap != null) {
        if (gson == null) {
          gson = new Gson();
        }
        Type gsonType = new TypeToken<HashMap>(){}.getType();
        String gsonString = gson.toJson(identifyMap,gsonType);
        Smartlook.setUserIdentifier(identifyKey, gsonString);
      } else if (identifyKey != null) {
        Smartlook.setUserIdentifier(identifyKey);
      }
    } else {
            result.notImplemented();
    }
  }
}

File: ios/Classes/SmartlookPlugin.m
#import "SmartlookPlugin.h"
#import <Smartlook/Smartlook.h>

#ifdef DEBUG
//#   define DLog(fmt, ...) NSLog((@"👀Smartlook: [Line %d] " fmt), __LINE__, ##__VA_ARGS__);
#   define DLog(...)
#else
#   define DLog(...)
#endif

#define IS_STRING_WITH_VALUE(string) ([string isKindOfClass:[NSString class]] && string > 0)
#define BOOL_VALUE_OR_DEFAULT(value,dft) ([value respondsToSelector:@selector(boolValue)] ? [value boolValue] : dft)
#define INT_VALUE_OR_DEFAULT(value,dft) ([value respondsToSelector:@selector(integerValue)] ? [value integerValue] : dft)

@implementation SmartlookPlugin

+ (void)registerWithRegistrar:(NSObject<FlutterPluginRegistrar>*)registrar {
  FlutterMethodChannel* channel = [FlutterMethodChannel
      methodChannelWithName:@"smartlook"
            binaryMessenger:[registrar messenger]];
  SmartlookPlugin* instance = [[SmartlookPlugin alloc] init];
  [registrar addMethodCallDelegate:instance channel:channel];
}

- (void)handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
    dispatch_async(dispatch_get_main_queue(), ^{
        [self _internal_handleMethodCall:call result:result];
    });
}

- (void)_internal_handleMethodCall:(FlutterMethodCall*)call result:(FlutterResult)result {
	
    DLog(@"Smartlook: flutter log:\nmethod: %@,\nstart args: %@", call.method, [call.arguments description]);

    if ([@"getPlatformVersion" isEqualToString:call.method]) {
		result([@"iOS lul" stringByAppendingString:[[UIDevice currentDevice] systemVersion]]);
        return;
    };

    // MARK: - API 1.1.0
    
    // MARK: Setup
    if ([@"setup" isEqualToString:call.method] || [@"setupAndStartRecording" isEqualToString:call.method]) {
        DLog(@"-> %@", call.method);
        NSString *key = call.arguments[@"key"];
        if (IS_STRING_WITH_VALUE(key)) {
            NSMutableDictionary *options = [NSMutableDictionary new];
            options[@"_forceRegularFramerate"] = @YES;
            NSNumber *fps = call.arguments[@"fps"];
            if ([fps isKindOfClass:[NSNumber class]]) {
                options[@"fps"] = fps;
            }
            [Smartlook setupWithKey:key options:options];
            if ([@"setupAndStartRecording" isEqualToString:call.method]) {
                [Smartlook startRecording];
            }
            [Smartlook setSessionPropertyValue:@"flutter" forName:@"sdk_build_flavor" withOptions:SLPropertyOptionImmutable];
        } else {
            DLog(@"Smartlook API key missing in %@ call", call.method);
        }
        return;
    }
    
    if ([@"setUserIdentifier" isEqualToString:call.method]) {
        DLog(@"-> %@", call.method);
        NSString *userIdentifier = call.arguments[@"key"];
        if (IS_STRING_WITH_VALUE(userIdentifier)) {
            [Smartlook setUserIdentifier:userIdentifier];
        }
        NSDictionary *map = call.arguments[@"map"];
        if ([map respondsToSelector:@selector(enumerateKeysAndObjectsUsingBlock:)]) {
            [map enumerateKeysAndObjectsUsingBlock:^(NSString * _Nonnull name, id _Nonnull value, BOOL * _Nonnull stop) {
                [Smartlook setSessionPropertyValue:value forName:name];
            }];
        }
        return;
    }
    
    // MARK: Recording
    if ([@"startRecording" isEqualToString:call.method]) {
        DLog(@"-> %@", call.method);
        [Smartlook startRecording];
        return;
    }
    if ([@"stopRecording" isEqualToString:call.method]) {
        DLog(@"-> %@", call.method);
        [Smartlook startRecording];
        return;
    }
    if ([@"isRecording" isEqualToString:call.method]) {
        DLog(@"-> %@", call.method);
        result([NSNumber numberWithBool:[Smartlook isRecording]]);
        return;
    }
    
    
    // MARK: Custom Events
    if ([@"startTimedCustomEvent" isEqualToString:call.method]) {
        DLog(@"-> %@", call.method);
        NSString *key = call.arguments[@"key"];
        if (IS_STRING_WITH_VALUE(key)) {
            [Smartlook startTimedCustomEventWithName:key props:@{}];
        }
        return;
    }
    if ([@"trackCustomEvent" isEqualToString:call.method]) {
        DLog(@"-> %@", call.method);
        NSString *key = call.arguments[@"key"];
        if (IS_STRING_WITH_VALUE(key)) {
            NSDictionary *map = [call.arguments[@"map"] respondsToSelector:@selector(enumerateKeysAndObjectsUsingBlock:)] ? call.arguments[@"map"] : @{};
            [Smartlook trackCustomEventWithName:key props:map];
        }
        return;
    }

    if ([@"trackNavigationEvent" isEqualToString:call.method]) {
        DLog(@"-> %@", call.method);
        NSString *key = call.arguments[@"key"];
        if (IS_STRING_WITH_VALUE(key)) {
            SLNavigationType navType = INT_VALUE_OR_DEFAULT(call.arguments[@"type"], 0) == 1 ? SLNavigationTypeExit : SLNavigationTypeEnter;
            [Smartlook tractNavigationEventWithControllerId:key type:navType];
        }
        return;
    }

    
    // MARK: Sensitive
    if ([@"startFullscreenSensitiveMode" isEqualToString:call.method]) {
        DLog(@"-> %@", call.method);
        [Smartlook beginFullscreenSensitiveMode];
        return;
    }
    if ([@"stopFullscreenSensitiveMode" isEqualToString:call.method]) {
        DLog(@"-> %@", call.method);
        [Smartlook endFullscreenSensitiveMode];
        return;
    }
    if ([@"isFullscreenSensitiveModeActive" isEqualToString:call.method]) {
        DLog(@"-> %@", call.method);
        result([NSNumber numberWithBool:[Smartlook isFullscreenSensitiveModeActive]]);
        return;
    }
    if ([@"enableWebviewRecording" isEqualToString:call.method]) {
        DLog(@"-> %@", call.method);
        BOOL enabled = BOOL_VALUE_OR_DEFAULT(call.arguments[@"enabled"], 0);
        if (enabled) {
            [Smartlook registerWhitelistedObject:[UIWebView class]];
            [Smartlook registerWhitelistedObject:[WKWebView class]];
        } else {
            [Smartlook registerBlacklistedObject:[UIWebView class]];
            [Smartlook registerBlacklistedObject:[WKWebView class]];
        }
        return;
    }
    
    
    // MARK: Global Properties
    if ([@"setGlobalEventProperty" isEqualToString:call.method]) {
        DLog(@"-> %@", call.method);
        NSString *name = call.arguments[@"key"];
        NSString *value = call.arguments[@"value"];
        if (!IS_STRING_WITH_VALUE(name) || !IS_STRING_WITH_VALUE(value)) {
            return;
        }
        SLPropertyOption option = BOOL_VALUE_OR_DEFAULT(call.arguments[@"immutable"], NO) ? SLPropertyOptionImmutable : SLPropertyOptionDefaults;
        [Smartlook setGlobalEventPropertyValue:value forName:name withOptions:option];
        return;
    }
    if ([@"setGlobalEventProperties" isEqualToString:call.method]) {
        DLog(@"-> %@", call.method);
        NSDictionary *map = call.arguments[@"map"];
        if ([map respondsToSelector:@selector(enumerateKeysAndObjectsUsingBlock:)]) {
            SLPropertyOption option = BOOL_VALUE_OR_DEFAULT(call.arguments[@"immutable"], NO) ? SLPropertyOptionImmutable : SLPropertyOptionDefaults;
            [map enumerateKeysAndObjectsUsingBlock:^(NSString * _Nonnull name, id _Nonnull value, BOOL * _Nonnull stop) {
               [Smartlook setGlobalEventPropertyValue:value forName:name withOptions:option];
            }];
        };
        return;
    }
    if ([@"removeGlobalEventProperty" isEqualToString:call.method]) {
        DLog(@"-> %@", call.method);
        NSString *key = call.arguments[@"key"];
        if (IS_STRING_WITH_VALUE(key)) {
            [Smartlook removeGlobalEventPropertyForName:key];
        }
        return;
    }
    if ([@"removeAllGlobalEventProperties" isEqualToString:call.method]) {
        DLog(@"-> %@", call.method);
        [Smartlook clearGlobalEventProperties];
        return;
    }
    
    
    // MARK: Others
    if ([@"getDashboardSessionUrl" isEqualToString:call.method]) {
        DLog(@"-> %@", call.method);
        result([Smartlook getDashboardSessionURL].absoluteString);
        return;
    }
    
    // MARK: Not available on iOS
    if ([@"setReferrer" isEqualToString:call.method]) {
        // not available on iOS
        DLog(@"-> %@", call.method);
        return;
    }
    if ([@"flush" isEqualToString:call.method]) {
        // not available on iOS
        DLog(@"-> %@", call.method);
        return;
    }
    if ([@"enableCrashlytics" isEqualToString:call.method]) {
        // not available on iOS
        DLog(@"-> %@", call.method);
        return;
    }


    
//
//
//
//
    //MARK:- Legacy code (to be removed later)
    
    // Start
    if ([@"start" isEqualToString:call.method]) {
        DLog(@"-> %@", call.method);
        DLog(@"flutter log: start args: %@, class: %@", [call.arguments description], [call.arguments class]);
        NSArray* args = call.arguments;
        if (args && args.count > 0) {
            NSString* key = [args objectAtIndex:0];
            if (key && [key isKindOfClass:[NSString class]] && key.length > 0) {
                [Smartlook setupWithKey:key options:@{ @"_forceRegularFramerate" : @YES}];
                [Smartlook startRecording];
            } else {
                DLog(@"Smartlook - missing API key");
            }
        } else {
            DLog(@"Smartlook - missing API key");
        }
        return;
    }
    if ([@"recordEvent" isEqualToString:call.method]) {
        DLog(@"-> %@", call.method);
        NSArray* args = call.arguments;
        if (args && args.count > 0) {
            NSString* eventName = [args objectAtIndex:0];
            
            if (eventName && [eventName isKindOfClass:[NSString class]] && eventName.length > 0) {
                [Smartlook recordCustomEventWithEventName:eventName propertiesDictionary:nil];
            } else {
                DLog(@"Smartlook - missing event name");
            }
        } else {
            DLog(@"Smartlook - missing event name");
        }
        return;
    }
    
    // MARK: Default
    DLog(@"Smartlook: method not found -> %@", call.method);
    result(FlutterMethodNotImplemented);
        
}

@end
